### ISSUE TITLE
ServicePartitionResolver.ResolveAsync hangs on internal COM exception

### ISSUE THREAD

**Issue by framgeld** [Azure/service-fabric-issues#874]
Although I have a cluster on the URI I am connecting to, the **ResolveAsync** method hangs (even though I provide a lot of timeouts).

Using the following code:

```
var task = serviceResolver.ResolveAsync(
    new Uri("fabric:/MyApp/MyService"),
    ServicePartitionKey.Singleton,
    TimeSpan.FromSeconds(15),
    TimeSpan.FromSeconds(15),
    new CancellationTokenSource(15000).Token);
var partition = task.GetAwaiter().GetResult();
```

I had to turn off "Just My Code" in the debugger to get this exception:

> Exception from HRESULT: 0x80071BBF
>    at System.Fabric.Interop.NativeClient.IFabricServiceManagementClient6.EndResolveServicePartition(IFabricAsyncOperationContext context)
>    at System.Fabric.FabricClient.ServiceManagementClient.ResolveServicePartitionEndWrapper(IFabricAsyncOperationContext context)
>    at System.Fabric.Interop.AsyncCallOutAdapter2`1.Finish(IFabricAsyncOperationContext context, Boolean expectedCompletedSynchronously)

Please advise how to fix this. I have been trying to get my WCF client working for 7 days now.

Thanks.

**Comment by oanapl** [comment]
The HRESULT you are seeing, 0x80071BBF is FABRIC_E_INVALID_PARTITION_KEY. 
In your call you pass ServicePartitionKey.Singleton. The [partition key](https://docs.microsoft.com/en-us/dotnet/api/microsoft.servicefabric.services.client.servicepartitionkey?view=azure-dotnet) you provide must have [Kind](https://docs.microsoft.com/en-us/dotnet/api/system.fabric.servicepartitionkind?view=azure-dotnet) matching the partitioning scheme used to create the service instance. What partition scheme are you using for your service? For example, if you are using UniformInt64 partition scheme in the service, pass a partition key that takes Int64 instead of singleton, even if there is just one partition that covers the entire key range.

**Comment by framgeld** [comment]
Thank you for the info. I have now changed my code so it uses this instead (as I am talking to a WCF listener on a stateful service with a single partition):

`new ServicePartitionKey(1)`

Unfortunately, the **ResolveAsync** method is still hanging indefinitely, and unlike the non-exposed COM error I was able to obtain last time, this time... nothing at all.

Getting the **ResolveAsync** method to work standalone was only for diagnostic purposes; to prove that it was the reason why the following code hangs (a point that I have proven to myself):

```
var clientFactory =
    new WcfCommunicationClientFactory<IProducerWcfService>(
        CreateBasicHttpBinding(),
        servicePartitionResolver: serviceResolver);

BasicHttpBinding CreateBasicHttpBinding()
{
    const int OneMegabyteInBytes = 1024 * 1024;

    return
        new BasicHttpBinding
        {
            SendTimeout = TimeSpan.MaxValue,
            ReceiveTimeout = TimeSpan.MaxValue,
            OpenTimeout = TimeSpan.FromSeconds(5),
            CloseTimeout = TimeSpan.FromSeconds(5),
            MaxReceivedMessageSize = OneMegabyteInBytes,
            MaxBufferSize = OneMegabyteInBytes,
            MaxBufferPoolSize = OneMegabyteInBytes * Environment.ProcessorCount,
        };
}

```

If you are aware of anything that can make it hang like this (no matter how _out there_ the suggestion may seem), I would be grateful if you can share. I am seriously out of ideas (and time); I am on my 8th working day trying to get this working and I am seriously flummoxed.

I should also mention that this is connecting to 3-node on-premise cluster, and is configured as an insecure cluster (which we will reinstall as a secure cluster once we have proven this connectivity works).

Thank you in advance for any further assistance you can provide.

P.S. When using the same code to connect to my local dev cluster, it works for one WCF call: the second WCF call hangs indefinitely.

**Comment by framgeld** [comment]
Another thing of note: my WCF client is server-side ASP.NET C# code (.NET 4.5). It is running in the debugger and is using 64-bit IIS Express.

**Comment by oanapl** [comment]
@BharatNarasimman, can you help?

**Comment by framgeld** [comment]
@oanapl I am currently discussing this with someone in Microsoft Support as we speak.

**Comment by framgeld** [comment]
@BharatNarasimman if you have any ideas what this could be, I would greatly appreciate it. No answers from Microsoft Support regarding this at the moment. A "two pronged approach" from different viewpoints may get this solved.

Thank you.

**Comment by framgeld** [comment]
After working with Microsoft Support and following some advice the engineer provided, I enabled WCF tracing on the Service Fabric WCF client. The log contained messages for the following (in chronological order):

- ServiceLevelSendRequest
- TransportSend
- TransportReceive
- ServiceLevelReceiveReply.

The second call (in which I believe ServicePartitionResolver.ResolveAsync is hanging), produced no client-side WCF traces whatsoever.

**Comment by BharatNarasimman** [comment]
@framgeld, [ServicePartitionResolver's code](https://github.com/Azure/service-fabric-services-and-actors-dotnet/blob/develop/src/Microsoft.ServiceFabric.Services/Client/ServicePartitionResolver.cs) throws operationcancelled after the cancellation token is signaled(15 seconds in your case)

Without using the wcfclient, can you try just calling resolveasync() - does using your code at the beginning of this thread, with this partitionkey - new ServicePartitionKey(1) hang for ever? Can you check if your service is showing a endpoint via SFX UI, also can you share the arguments you pass to create service, if you are not explicitly calling create service, can you share the <DefaultServices> section from your application manifest?

**Comment by framgeld** [comment]
@BharatNarasimman This github issue is a "child" of this larger issue that I created, which may shed more light on the issues/challenges I am facing:

https://github.com/Azure/service-fabric-issues/issues/867 Multiple issues with WCF communications to stateful service

Using the partition key you asked for and calling ResolveAsync... yes, it hangs forever - but only on the second attempt (the same as the WCF call that I try to make). Below is the code I used to test this.

```
string[] serviceFabricClusterAddresses =
{
    "localhost:19000",
    "localhost:19001",
};

var serviceResolver = new ServicePartitionResolver(serviceFabricClusterAddresses);
Task<ResolvedServicePartition> task =
    serviceResolver.ResolveAsync(
        "fabric:/MyApp/MyService",
        new ServicePartitionKey(1),
        TimeSpan.FromSeconds(15),
        TimeSpan.FromSeconds(15),
        new CancellationTokenSource(15000).Token);
ResolvedServicePartition resolvedPartition = task.GetAwaiter().GetResult();
```

Service Fabric Explorer shows the following endpoints against my primary replica, and shows no endpoints against my secondary replicas:

- External Wcf Service Endpoint | http://localhost:31001/
- V 2 Listener | localhost:1818+5cbdd14a-804f-4da7-8735-203793d8e46c-131641992214048682-db08700a-07c7-4ae2-a1b8-2e32d737ad6a

The ServiceManifest.xml contains the following configuration for the endpoints:

```
<Endpoints>
  <Endpoint Name="ExternalWcfServiceEndpoint" Protocol="http" UriScheme="http" />
  <Endpoint Name="V2Listener" />
  <Endpoint Name="ReplicatorEndpoint" />
</Endpoints>
```

Below is the code used to create the service (IProducerWcfService is for my external WCF listener, while IProducerRemotingService is for the ServiceProxy v2 listener).

```
namespace <removed>
{
    <proprietary using statements removed>
    using System;
    using System.Collections.Generic;
    using System.Fabric;
    using System.ServiceModel;
    using System.ServiceModel.Description;
    using System.Threading;
    using System.Threading.Tasks;

    using Microsoft.ServiceFabric.Services.Communication.Runtime;
    using Microsoft.ServiceFabric.Services.Communication.Wcf.Runtime;
    using Microsoft.ServiceFabric.Services.Remoting.Runtime;
    using Microsoft.ServiceFabric.Services.Runtime;

    public sealed class Microservice
        : StatefulService, IProducerRemotingService
    {
        public Microservice(StatefulServiceContext context)
            : base(context)
        {
        }

        internal JobScheduler JobScheduler { get; private set; }

        internal Queues Queues { get; private set; }

        internal CancellationToken CancellationToken { get; private set; }

        public Task<JobBase> GetNextJob()
        {
            return Task.Run(
                () => DequeueJob(),
                this.CancellationToken);

            JobBase DequeueJob()
            {
                return this.Queues?.DequeueJob();
            }
        }

        public Task<bool> AcknowledgeReceiptOfJob(
            JobBase job)
        {
            return Task.Run(
                () => AcknowledgeReceipt(),
                this.CancellationToken);

            bool AcknowledgeReceipt()
            {
                return this.Queues?.AcknowledgeReceiptOfDequeuedJob(job) == true;
            }
        }

        public Task<bool> ConfirmJobStillInProgress(
            JobBase job)
        {
            return Task.Run(
                () => ConfirmStillInProgress(),
                this.CancellationToken);

            bool ConfirmStillInProgress()
            {
                return this.Queues?.ConfirmDequeuedJobIsStillInProgress(job) == true;
            }
        }

        public Task<bool> MarkJobAsComplete(
            JobBase job)
        {
            return Task.Run(
                () => MarkAsComplete(),
                this.CancellationToken);

            bool MarkAsComplete()
            {
                return this.Queues?.MarkJobAsComplete(job) == true;
            }
        }

        protected override IEnumerable<ServiceReplicaListener> CreateServiceReplicaListeners()
        {
            // Due to an issue with Service Fabric and WCF clients that connect to the service,
            // this listener MUST be created first. The WCF client seems to connect to the first
            // listener when connecting from outside the cluster.
            yield return CreateProducerWcfServiceListener();

            foreach (var listener in this.CreateServiceRemotingReplicaListeners())
            {
                yield return listener;
            }

            ServiceReplicaListener CreateProducerWcfServiceListener()
            {
                const string ExternalWcfServiceEndpointResourceName = "ExternalWcfServiceEndpoint";
                return
                    new ServiceReplicaListener(
                        GetWcfCommunicationListener,
                        ExternalWcfServiceEndpointResourceName);

                WcfCommunicationListener<IProducerWcfService> GetWcfCommunicationListener(
                    StatefulServiceContext context)
                {
                    string uri = GetUri();
                    return
                        new WcfCommunicationListener<IProducerWcfService>(
                            context,
                            new ProducerWcfService(this),
                            new BasicHttpBinding(BasicHttpSecurityMode.None),
                            GetEndpointAddress());

                    string GetUri()
                    {
                        var endpointConfig = context.CodePackageActivationContext.GetEndpoint(ExternalWcfServiceEndpointResourceName);
                        int port = endpointConfig.Port;
                        string scheme = endpointConfig.Protocol.ToString();
                        string host = context.NodeContext.IPAddressOrFQDN;
                        return $"{scheme}://{host}:{port.ToInvariantString()}/";
                    }

                    EndpointAddress GetEndpointAddress()
                    {
                        return new EndpointAddress(uri);
                    }
                }
            }
        }

        protected override async Task RunAsync(CancellationToken cancellationToken)
        {
            this.CancellationToken = cancellationToken;

            await Task.Run(() => StartService(), cancellationToken);
            await PauseUntilServiceIsShutDown();
            StopService();

            void StartService()
            {
                this.Queues = new Queues(this);
                this.JobScheduler = new JobScheduler(this);
                this.JobScheduler.RefreshScheduler();
                this.JobScheduler.StartScheduler();
            }

            async Task PauseUntilServiceIsShutDown()
            {
                await Task.Delay(Timeout.Infinite, cancellationToken);
            }

            void StopService()
            {
                try
                {
                    this.JobScheduler?.StopScheduler();
                    SystemEventLogging.Current?.StopSystemEventLogging();
                }
                catch (Exception)
                {
                    // Ignore errors raised during shutdown.
                    // It is possible these are due to the cancellation token being
                    // set, and therefore the service never got to start up properly.
                }
            }
        }
    }
}
```

I also tried to make my stateful service implement my IProducerWcfService (in case having that separated into its own class was causing the issue), but I still had the same issues.

Thank you.

**Comment by framgeld** [comment]
@BharatNarasimman I'm grasping at straws in case I can find something of use to help us to figure this out... maybe this has something to do with it...?

You will notice in my Stateful Service code from my previous post, that there is a comment with regards to maintaining the order in which my service creates the listeners. This is because when the V2 ServiceProxy listener was created first, the Service Fabric WCF client was throwing an error along the lines of "localhost scheme invalid, should be http". I narrowed it down to the way in which the Service Fabric WCF client resolves its endpoints.

The **ResolvedServicePartition.Endpoints** property appears to contain a collection of endpoints with the count being equal to the number of Stateful Service replicas; not the number of endpoints on the primary replica. The WCF client then uses the first endpoint against the primary replica, which causes issues when the first endpoint is the V2 ServiceProxy listener.

Below is the contents of the first item in the **ResolvedServicePartition.Endpoints** property:

`{"Endpoints":{"ExternalWcfServiceEndpoint":"http:\/\/localhost:30001\/",
"V2Listener":"localhost:2261+a26750f1-5c98-42b6-bfff-6a90a6f412f1-131642793450142627-38a8059a-2b55-42d7-bc3f-71b5ae7f6668"}}`

**Comment by framgeld** [comment]
@BharatNarasimman I performed another test that may assist with your investigation.

I created two do/while loops:

- **Loop 1** - Create **ServicePartitionResolver** once (outside the loop), and call **ResolveAsync** in each loop cycle.
- **Loop 2** - Re-create **ServicePartitionResolver** for each loop cycle and call **ResolveAsync** on it.

**Loop 1** never hangs, while **Loop 2** hangs on the second loop cycle. Is **Loop 1** working because **ResolveAsync** uses a cached version, or is this a clue?

**Comment by framgeld** [comment]
@BharatNarasimman This issue can now be closed. For others reading this and experiencing the same issue:

This was related to async programming. The WCF client's WCF method was attempting to run asynchronously on the same ASP.NET worker thread that spawned the async Task for the WCF method call. As a result, it was causing deadlocks. I ensured my API "all the way down" was changed to use async / await, and the library await statements were changed to use ConfigureAwait(false).
